client/doc.go

```diff
@@ -0,0 +1,156 @@
+(1) The client attempts to register a name that already exists in the
+CONIKS key directory (ErrorNameExisted); the server returns a
+privacy-preserving proof of inclusion in its response. (2) The server
+encounters an internal error when attempting to register the name
+(ErrorDirectory).
+
+Assumptions and Possible Attacks: The client accepts the STR in the
+registration proof and assumes that the server has presented it with a
+legitimate history with this initial STR (i.e. Tofu).
+The main downside to the Tofu approach is that this does
+allow a malicious server to place the client
+on a forked branch of its history. However, by auditing the initial STR,
+the client can detect this attack. Alternatively, the client could verify
+the server's entire prior history to ensure that the server has not attempted
+to equivocate about its initial STR. This approach would be more secure,
+but has performance penalties. For now, we believe auditing will mitigate
+this attack, but the need for a full prior history check upon registration
+remains under discussion.
+
+As mentinoed above, for performance reasons, the client does not check
+the server's prior hisotry upon registering a new name. This enables
+a malicious server to remove a legitimate username from the directory
+and allow another user to re-register the name as the registering
+client would not discover the prior existence of the name.
+The server is then able to deny service to the original user by allowing
+another user to re-register the name and use the name. In the worst case,
+the server is complicit in a malicious user's attempt to assume the
+identity of the legitimate user. However, if the server has created a
+fork in its history to equivocate and attempt to conceal this attack,
+the client of the original owner of the stolen username will detect
+this attack via auditing and monitoring, and the registering client
+will detect this attack during the auditing step of the registration
+protocol.
+
+Not checking the server's prior history during the registration protocol
+also allows a malicious server to register a username, and remove
+this name from the directory when a legitimate user registers the same
+name to allow for the registration request to succeed. Assuming the server
+has created a fork in its history at the point of the legitimate
+registration, it can attempt to equivocate about the name. However, the
+registering client can detect this
+attack during the auditing step of the registration protocol.
+
+What these attacks have in common is that they take advantage
+of the fact that the registration protocol does not require the client
+to verify the server's prior history allowing an illegitimate
+re-registration of a username.  As a countermeasure, CONIKS does not
+currently allow re-registration of retired usernames,
+i.e. the user has deactivated and forfeit her account thereby making her
+username available for use by another user. The design decision is
```

Comment:
Right, not allowing re-registrations is by policy, but the append-only property of the tree is by design, though. You're right, there's no proof that a name was registered and then removed, so verifying the entire prior history upon registering a new name would mitigate this issue.